### PR TITLE
Features/62 resize field

### LINKED ISSUE #62: Resize field dynamically according to player count, columns and queues

### PR DESCRIPTION
**Pull request by arekisanda**
Closes #62

### CODE REVIEW

**Comment by arekisanda** [comment]
# Basic logic
So far, I was able to calculate the resizing operations (add/remove columns/queues).

# Integration
I tried to integrate the resizing logic in `GameServiceImpl` and left some `TODO`s for optional performance improvements and event emissions.

# Testing
I was not able to test the integration locally, as `docker-compose` does not allow the `extends` keyword in version 3.3.

**Review by mjossdev — APPROVED** [review]

File: backend/rohan/src/main/kotlin/ch/chaosconnect/rohan/services/GameServiceImpl.kt [hunk 1]
```diff
@@ -215,14 +218,86 @@ class GameServiceImpl(private val storageService: StorageService) :
...
+            val (queuesHeadSuggestion, queuesTailSuggestion) = calculateCollectionResizingSuggestions(
+                queues,
+                targetColumnCount
+            ) {
+                it.isNotEmpty()
+            }
+            val headSuggestion =
+                max(columnsHeadSuggestion, queuesHeadSuggestion)
+            val tailSuggestion =
+                max(columnsTailSuggestion, queuesTailSuggestion)
```
[the remark this answers is not in the document]
Comment (arekisanda) [reply backend/rohan/src/main/kotlin/ch/chaosconnect/rohan/services/GameServiceImpl.kt:240]: The size should indeed be always the same. I'm more worried about different cell counts.

# Example:

`targetColumnCount`: 2 (not necessarily realistic)

`columns`:
```       
  Y    
 YRR   
=======
1   321
h   t
```       

`columnsHeadSuggestion`: -1

`columnsTailSuggestion`: -3

`queues`:
```       
  y r   
=======
12   21
 h   t
```       

`queuesHeadSuggestion` -2

`queuesTailSuggestion`: -2

`columns` + `queues`:
```
  y    
  Y    
 YRRr  
=======
1    21
h    t
```       

`headSuggestion`: -1
`tailSuggestion`: -2

[the remark this answers is not in the document]
Comment (PascalHonegger) [reply backend/rohan/src/main/kotlin/ch/chaosconnect/rohan/services/GameServiceImpl.kt:240]: As far as I understood the coed the column and queue are always added and reduced at the same time so this scenario shouldn't be possible (Hopefully 😉).

[the remark this answers is not in the document]
Comment (arekisanda) [reply backend/rohan/src/main/kotlin/ch/chaosconnect/rohan/services/GameServiceImpl.kt:240]: Once #67 is merged as well, we can repace
```
            val (columnsHeadSuggestion, columnsTailSuggestion) = calculateCollectionResizingSuggestions(
                columns,
                targetColumnCount
            ) {
                it.isNotEmpty()
            }
            val (queuesHeadSuggestion, queuesTailSuggestion) = calculateCollectionResizingSuggestions(
                queues,
                targetColumnCount
            ) {
                it.isNotEmpty()
            }
            val headSuggestion =
                max(columnsHeadSuggestion, queuesHeadSuggestion)
            val tailSuggestion =
                max(columnsTailSuggestion, queuesTailSuggestion)
```
in GameServiceImpl.kt with
```
            val (headSuggestion, tailSuggestion) = calculateCollectionResizingSuggestions(
                columns,
                targetColumnCount
            ) {
                it.rows.isNotEmpty() || it.hasQueue || !it.isDisabled
            }
```
.

**Pull request merged**